tutorial/openmp.md

```diff
@@ -0,0 +1,573 @@
+```python
+from numpy import zeros
+result = 0
+n = 1337
+arr = zeros(n, dtype=int)
+#$ omp parallel num_threads(4)
+#$ omp simd
+for i in range(0, n):
+  arr[i] = i
+#$ omp end parallel
+for i in range(0, n):
+  result = result + arr[i]
+print("Result:", result)
+```
+
+The output of this program is:
+```shell
+❯ pyccel omp_test.py --language c --openmp
+❯ ./omp_test
+Result: 893116
+```
+
+### Task / Taskwait Construct
+
+#### Syntax Task Construct
+
+```python
+#$ omp task [clause[ [,]clause] ... ]
+structured-block
+#$ omp end task
+```
+
+#### Syntax Taskwait Construct
+
+```python
+#$ omp taskwait
+```
+
+#### Example
+
+The ``` #$ omp task ``` pragma is used here to define an explicit task.\
+The ``` #$ omp taskwait ``` pragma is used here to specify a wait on the completion of child tasks of the current task.
+```python
+@types('int', results='int')
+def fib(n):
+  if n < 2:
+    return n
+  #$ omp task shared(i) firstprivate(n)
+  i = fib(n-1)
+  #$ omp end task
+  #$ omp task shared(j) firstprivate(n)
+  j = fib(n-2)
+  #$ omp end task
+  #$ omp taskwait
+  return i+j
+
+#$ omp parallel
+#$ omp omp single
+print(fib(10))
+#$ omp end single
+#$ omp end parallel
+
+print("result :", result)
+```
+
+The output of this program is:
+```shell
+❯ pyccel omp_test.py --language c --openmp
+❯ ./omp_test
+55
+```
+
+### Taskyield Construct
+
+#### Syntax
+
+```python
+#$ omp taskyield
+```
+
+#### Example
+
+The ``` #$ omp taskyield ``` pragma specifies that the current task can be suspended at this point, in favor of execution of a different task.
+
+```python
+#$ omp task
+long_function()
+#$ omp taskyield
+long_function_2()
+#$ omp end task
+```
+
+### Flush Construct
+
+#### Syntax
+
+```python
+#$ omp flush
+```
+
+#### Example
+
+The ``` #$ omp flush ``` pragma is used to ensure that all threads in a team have a consistent view of certain objects in memory.
+```python
+from pyccel.stdlib.internal.openmp import omp_get_thread_num
+flag = 0
+#$ omp parallel num_threads(2)
+if omp_get_thread_num() == 0:
+  #$ omp atomic update
+  flag = flag + 1
+elif omp_get_thread_num() == 1:
+  #$ omp flush(flag)
+  while flag < 1:
+    #$ omp flush(flag)
+  print("Thread 1 released")
+  #$ omp atomic update
+  flag = flag + 1
+#$ omp end parallel
+print("flag:", flag)
+```
+
+The output of this program is:
+```shell
+❯ pyccel omp_test.py --language c --openmp
+❯ ./omp_test
+Thread 1 released
+flag: 2
+```
+
+### Cancel Construct
+
+#### Syntax
+
+```python
+#$ omp cancel construct-type-clause[ [ , ] if-clause]
+```
+
+#### Example
+
+The ``` #$ omp cancel ``` is used to request cancellation of the innermost enclosing region of the type specified.
+```python
+result = 0
+#$ omp parallel
+#$ omp for private(i) reduction (+:result)
+for i in range(len(v)):
+  result = result + v[i]
+  if result < 0:
+    #$ omp cancel for
```

Comment:
yes `cancel`  is equivalent to a break, we need to add `break`.
i just checked the break statement not implemented yet for C.

Reply:
Yes I don't think `continue` is implemented either. There are tests that could be activated in `epyccel/test_loops.py` if it was fixed. It would be a very quick fix. But if you leave the default langugae then it's not a huge problem as they are implemented in fortran

Reply:
yes, so im gonna leave `continue` and `break ` to another pull request.